src/pages/index.js

```diff
@@ -24,7 +40,416 @@ const profileModalTitleInput =
+  addCardModal.setLoading(true);
+  api
+    .createNewCard({ name: addCardData.title, link: addCardData.url })
+    .then((cardData) => {
+      createCard(cardData);
+      addCardModal.close();
+    })
+    .catch((err) => console.error("Error adding new card:", err))
+    .finally(() => {
+      addCardModal.setLoading(false);
+    });
+});
+addCardModal.setEventListeners();
+*/
+
+const profileModal = new ModalWithForm("#profile-edit-modal", (profileData) => {
+  profileModal.renderLoading(true);
+  handleProfileModalSubmit(profileData);
+});
+
+profileModal.setEventListeners();
+/* Previous profileModal
+const profileModal = new ModalWithForm("#profile-edit-modal", (profileData) => {
+  profileModal.setLoading(true);
+
+  api
+    .updateProfileInfo(profileData.title, profileData.description)
+    .then((updatedUserData) => {
+      userInfo.setUserInfo(updatedUserData.title, updatedUserData.description);
+      profileModal.close();
+    })
+    .catch((err) =>
+      console.error("Error updating profile user information:", err)
+    )
+    .finally(() => {
+      profileModal.setLoading(false);
+    });
+});
+profileModal.setEventListeners();
+*/
+
+//
+// ModalWithImage Instantiation
+//
+const imagePreviewModal = new ModalWithImage("#preview-image-modal");
+imagePreviewModal.setEventListeners();
+
+//
+// Profile Update Avatar Modal Instantiation
+//
+const avatarUpdateModal = new ModalWithForm(
+  "#profile-avatar-modal",
+  (avatarData) => {
+    avatarUpdateModal.renderLoading(true);
+
+    api
+      .updateAvatar(avatarData.url)
+      .then((res) => {
+        userInfo.changeAvatar(res.avatar);
+        avatarUpdateModal.close();
+      })
+      .catch((err) => console.error("Error updating profile image:", err))
+      .finally(() => {
+        avatarUpdateModal.renderLoading(false);
+      });
+  }
+);
+avatarUpdateModal.setEventListeners();
+
+//
+// ModalConfirmDelete Instantiation
+//
+const deleteConfirmation = new ModalConfirmDelete("#delete-card-confirm-modal");
+deleteConfirmation.setEventListeners();
+
+//
+// Form Validation
+//
+const formValidators = {};
+
+const enableValidation = (config) => {
+  const formList = Array.from(document.querySelectorAll(".modal__form"));
+
+  formList.forEach((formElement) => {
+    const validator = new FormValidator(config, formElement);
+
+    const formName = formElement.getAttribute("name");
+
+    formValidators[formName] = validator;
+    validator.enableValidation();
+  });
+};
+
+enableValidation(configItems);
+
+//
+//Event Handlers
+//
+function handleProfileModalSubmit(data) {
+  api
+    .updateProfileInfo(data.title, data.description)
+    .then((userData) => {
+      userInfo.setUserInfo({
+        title: userData.name,
+        description: userData.about,
+      });
+      profileModal.close();
+    })
+    .catch((err) => console.error(`Failed to update users info: ${err}`))
+    .finally(() => {
+      profileModal.renderLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

Reply:
Hello, thank you for reviewing and approving my project. I see you left a Could be improved comment. I've attached a screenshot of what I believe you are suggesting I do. If you could look over it and see if I have the correct approach (I haven't pushed these changes, but everything in the code is working as it should). If my approach is not correct, then I will leave my code as it originally was (I don't want to take up more of your time). Again, thank you.
![Screenshot 2024-09-26 141347](https://github.com/user-attachments/assets/5b365784-c920-4c6b-812c-58b13aec81aa)